### PR TITLE
Startup issues

### PR DESCRIPTION
**Pull request by wenningerk**
Fixes the issue we recently discussed on the clusterlabs-list that sbd failing to start didn't prevent pacemaker from coming up.
This is kind of severe as pacemaker would run without supervision by sbd, watchdog would possibly not be armed and the poison-pill wouldn't be read from the storage.
What led to this finding was systemd timing out on trying to start sbd when SBD_DELAY_START is enabled and msgwait > default startup-timeout.
Thus I added some reminder to check for the timeout to sbd.sysconfig.

In the code - right before enabling the watchdog it says:
```
/* We only do this at the point either the disk or
 * cluster servants become healthy
 */
```
In case of having no disks configured this would mean that my patch would create a deadlock.
But fortunately this comment seems to be a leftover. In the case without disks it still does detach from
the main-process - enabling the watchdog in the course of this - but it doesn't trigger as long as
there is no cluster.
So everything should be fine so that once systemd detects proper start of sbd it should be assured that the watchdog could be opened.

### CODE REVIEW

**Comment by krig** [comment]
Wouldn't `RequiredBy=pacemaker.service` make `sbd` a requirement of pacemaker, then? I mean, it's still feasible to run Pacemaker without SBD, right?

**Comment by wenningerk** [comment]
Nope - that is the cool thing about the Install section that the dependency is just created if sbd is enabled!

**Comment by krig** [comment]
Aha, I see!

File: src/sbd.service.in [hunk 1]
```diff
@@ -22,4 +22,5 @@ Restart=on-abort
 
 [Install]
 RequiredBy=corosync.service
-
+RequiredBy=pacemaker.service
```
Comment (arvidjaar) [inline src/sbd.service.in]: This is wrong (it works, but it is big hammer) because

- `sbd.service` is already `RequiredBy corosync.service` and `corosync.service` is already `RequiredBy pacemaker.service` so adding extra `RequiredBy=pacemaker.service` to `sbd.service` is completely redundant.

- it does not fix the problem that `corosync.service` still starts even though `sbd.service` startup eventually fails. Because startup of `pacemaker.service` is delayed, what happens is - `corosync.service` is started, `sbd.service` startup fails, `pacemaker.service` startup is cancelled due to dependency failure, `corosync.service` is stopped again because it is not needed. The whole dance can be avoided by simply adding `Before=corosync.service` to `sbd.service` to avoid unnecessary starting of corosync in the first place.

File: src/sbd.service.in [hunk 2]
```diff
@@ -22,4 +22,5 @@ Restart=on-abort
...
+RequiredBy=dlm.service
```
Comment (arvidjaar) [inline src/sbd.service.in]: This is useless without also adding `Before=dlm.service`.

Comment (wenningerk) [reply src/sbd.service.in]: guess it is symmetric to the pacemaker-case ...

File: src/sbd.service.in (hunk 1 above)
Comment (wenningerk) [reply src/sbd.service.in]: Unfortunately for startup of sbd, in case of no disks configured, corosync has to be there.
That is what initially created the trouble when usage of sbd in watchdog-only mode was introduced. Before that it was easy like starting first sbd and corosync afterwards.
At least that is how I get it at the moment - have to try it again.
To a certain point the behavior with the configuration as is could be called a systemd bug as well as sbd is required by corosync and pacemaker requires corosync and pacemaker is to be started after corosync. Haven't checked how that is implemented and from what I've seen documentation is unclear about it. Seems as if this behavior comes from PartOf present as well.

What we could think of - as maybe a cleaner approach - would be the introduction of a target like  pre-pacemaker.target or cluster-layer.target.
Unfortunately to be really cleaner that would require synchronized adaptions in all 3 packages (sbd, corosync, pacemaker).

Comment (arvidjaar) [reply src/sbd.service.in]: @wenningerk 

This is exactly why I said on mailing list that we need clear understanding of relationship between services in plain English before we attempt to formalize them using systemd directives.

> the behavior you described under your 2nd bullet is exactly what we want

Good. Then remove `RequireBy=corosync.service` from `sbd.service` because this line implies exactly the opposite behavior. Moreover, if "for startup of sbd corosync has to be there" is also not ensured - corosync and sbd are started concurrently, and it is non-deterministic which will be launched first. You really need `After=corosync.service` in `sbd.service` then.

> the behavior with the configuration as is could be called a systemd bug ... documentation is unclear about it

That's true; what's worse, developers do not see what is wrong with documentation because they interpret it basing on their knowledge about implementation and cannot understand that others may interpret documentation differently.

OK, small explanation. For purpose of the following, "pending job" is job that was added to systemd queue but not yet selected for execution.

Systemd dependencies initially were about jobs and jobs only. They have nothing to do with unit states as users tend to interpret documentation. The only thing `A Requires B` does - if start job for B fails, **pending** start job for A is cancelled. For this to work there must be pending job for A in the first place, and the only possibility to ensure it is to say `A After B`. If `A After B` is missing, both start jobs for A and B a selected for execution concurrently, so even if job for B fails, it does not affect start job for A in any way (at this time A ay actually have long been started).

Which is why systemd had to introduce this hideous mongrel `BindsTo` which actually reacts to unit state change itself in addition to job dependency.

I have yet to see a single example where `A Requires B` is useful without `A After B`. So far every time I have seen it was in bug reports (due to this misconfiguration).

Comment (wenningerk) [reply src/sbd.service.in]: Ok, let me try some English ;-)
Please correct me if you have a different opinion or if I have missed something.

If startup of sbd needs corosync depends on the configuration of sbd.
In case that disks are configured with sbd there are at least parts that don't need corosync.
So it definitely makes sense to start them up in parallel so that either party can do the independent stuff and - if sbd config does require that - meet seeing each other.
Further more it shouldn't be possible to just start sbd without corosync or to trigger start of corosync by starting sbd.
Pacemaker should just be started if corosync is up and if enabled sbd has to be there as well.
And of course all the restarts of services you could possibly issue should lead to 
a behavior that makes sense in some way and after everything is settled again it should
work.

Back to how it is being told to systemd...
```RequiredBy=corosync.service``` for sbd might definitely be arguable if ```RequiredBy=sbd.service``` is anyway mandatory for pacemaker.
As ```PartOf``` allows the possibility that services can kind of depend on each other
it on the other hand doesn't hurt either.
And so you would as well be able to cover a case where corosync is being used by
something that actually isn't aware of fencing. (3 disks and if you don't see a quorate
number you would reboot) Or just as a test to see if everything below pacemaker
is working ...
Of course it would be good to assure that my current suggestion (that part is actually
inherited from what we have so far) doesn't misuse some kind of systemd behavior
that might change over time.
Thus I would suggest to go with my suggestion for now as it doesn't make things
worse regarding this issue and on the other hand solves the issue with pacemaker
being started on failing sbd and doesn't change legacy behavior people expect in
any other way.
On a longer term I'd suggest to go the route with an interimistic target pacemaker
depends on as - at least at first sight - it seems to use systemd capabilities in a
clearer way. 
Although that might not be so straight forward as we have to consider shutdown
and restart cases as well:
If corosync e.g. goes down considerably before sbd we would see a watchdog
reboot. Of course nothing we couldn't cover by sbd detecting a graceful shutdown
of corosync or just selection of proper timeouts ...

**Comment by wenningerk** [comment]
@Andrei: Your input is really appreciated but at the moment I don't see a way to do it 'properly' because of the dependencies to so many other packages like corosync, pacemaker and the highlevel-tooling (at least pcs) and maybe dlm
I'll go the route of pulling it as is (should basically solve the issue of having a cluster-node coming up without sbd-supervision even if configured - and doesn't make the situation worse in any way I'm aware of) and opening a discussion on the clusterlabs-list to discuss a cleaner and more future-proof approach.
